editoast/src/modelsv2/projects.rs

```diff
@@ -0,0 +1,265 @@
+    pub tags: Tags,
+    #[model(column = "image_id")]
+    pub image: Option<i64>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default, ToSchema, PartialEq)]
+pub struct Tags(Vec<String>);
+
+impl Tags {
+    pub fn new(value: Vec<String>) -> Self {
+        Self(value)
+    }
+}
+
+impl From<Vec<Option<String>>> for Tags {
+    fn from(value: Vec<Option<String>>) -> Self {
+        Self(value.into_iter().flatten().collect())
+    }
+}
+
+impl From<Tags> for Vec<Option<String>> {
+    fn from(value: Tags) -> Self {
+        value.0.into_iter().map(Some).collect()
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Default, ToSchema)]
+pub enum Ordering {
+    NameAsc,
+    NameDesc,
+    CreationDateAsc,
+    CreationDateDesc,
+    #[default]
+    LastModifiedDesc,
+    LastModifiedAsc,
+}
+
+impl Ordering {
+    pub fn to_sql(&self) -> &str {
+        match *self {
+            Ordering::NameAsc => "LOWER(t.name) ASC",
+            Ordering::NameDesc => " LOWER(t.name) DESC",
+            Ordering::CreationDateAsc => "creation_date",
+            Ordering::CreationDateDesc => "creation_date DESC",
+            Ordering::LastModifiedAsc => "last_modification",
+            Ordering::LastModifiedDesc => "last_modification DESC",
+        }
+    }
+}
+
+impl Project {
+    /// This function takes a filled project and update to now the last_modification field
+    pub async fn update_last_modified(self, db_pool: Data<DbPool>) -> Result<Option<Project>> {
+        let mut conn = db_pool.get().await?;
+        self.update_last_modified_conn(&mut conn).await
+    }
+
+    pub async fn update_last_modified_conn(
+        mut self,
+        conn: &mut PgConnection,
+    ) -> Result<Option<Project>> {
+        self.last_modification = Utc::now().naive_utc();
+        self.save(conn).await?;
+        Ok(Some(self))
+    }
+
+    pub async fn with_studies(self, db_pool: Data<DbPool>) -> Result<ProjectWithStudies> {
+        use crate::tables::study::dsl as study_dsl;
+        let conn = &mut db_pool.get().await?;
+        let studies_count = study_dsl::study
+            .filter(study_dsl::project_id.eq(self.id))
+            .count()
+            .get_result(conn)
+            .await?;
+        Ok(ProjectWithStudies::new_from_project(self, studies_count))
+    }
+
+    pub async fn delete_and_prune_document(
+        conn: &mut PgConnection,
+        project_id: i64,
+    ) -> Result<bool> {
+        let project_obj =
+            Project::retrieve_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
+                .await?;
+        let _ = Project::delete_static(conn, project_id).await?;
+
+        if let Some(image) = project_obj.image {
+            // We don't check the result. We don't want to throw an error if the image is used in another project.
+            let _ = Document::delete_static(conn, image).await;
+        }
+
+        Ok(true)
+    }
+}
+
+pub trait UpdateAndPruneDocument {
```

Comment:
because I got this error without the trait:
```
error[E0118]: no nominal type found for inherent implementation
   --> src/modelsv2/projects.rs:138:1
    |
138 | impl Changeset<Project> {
    | ^^^^^^^^^^^^^^^^^^^^^^^ impl requires a nominal type

Reply:
Right. Too bad rustc can't resolve straightforward refs like these ones :'(

I guess we don't have much choice then:

* `impl ProjectChangeset`
* `Project::update_and_prune_document(conn, cs: Changeset<Self>, id: i64)`

I have a slight preference for the latter as it doesn't uses generated names.

Reply:
Done.